### PR TITLE
EDGEPOD-372:Context Not found error handling for TAU, Service Request…

### PR DESCRIPTION
**Pull request by clstefi**
This activity involves the following changes:
1.Inclusion of componentDb (to eliminate the use of extern MmeIpcIf*)
2.Handling of UeCtxtNotFound error in DDN, Service Request and TAU
3.Context Manager Crash Fixes

### CODE REVIEW

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

File: src/mme-app/actionHandlers/attachActionHandlers.cpp [hunk 1]
```diff
@@ -646,6 +649,11 @@ ActStatus ActionHandlers::process_esm_info_resp(SM::ControlBlock& cb)
 
 	const s1_incoming_msg_data_t* s1_msg_data = static_cast<const s1_incoming_msg_data_t*>(msgBuf->getDataPointer());
 	const struct esm_resp_Q_msg &esm_res =s1_msg_data->msg_data.esm_resp_Q_msg_m;
+    
+    	if (esm_res.status != SUCCESS)
+    	{
+        	log_msg(LOG_ERROR, "ESM Response failed \n");
+    	}
 
 	procedure_p->setRequestedApn(Apn_name(esm_res.apn));
```
Comment (thakurajayL) [inline src/mme-app/actionHandlers/attachActionHandlers.cpp:658]: if ESM response failed then we should not use esm_res.apn.  You may wish to open a enhancement defect in MME to handle ESM response failure.

File: src/mme-app/actionHandlers/serviceRequestActionHandlers.cpp [hunk 2]
```diff
@@ -133,16 +134,24 @@ ActStatus ActionHandlers::send_ddn_ack_to_sgw(ControlBlock& cb)
 	    return ActStatus::HALT;
 	}
```
Comment (thakurajayL) [inline src/mme-app/actionHandlers/serviceRequestActionHandlers.cpp:136]: Open a enhancement in MME to send DDN Ack with -ve cause if no UE context found.

File: src/mme-app/msgHandlers/gtpMsgHandler.cpp [hunk 3]
```diff
@@ -170,10 +172,11 @@ void GtpMsgHandler::handleDdnMsg_v(IpcEventMessage* eMsg, uint32_t ueIdx)
 {
 	log_msg(LOG_INFO,"Inside handle DDN\n");
 
-	SM::ControlBlock* controlBlk_p = SubsDataGroupManager::Instance()->findControlBlock(ueIdx);
+	SM::ControlBlock* controlBlk_p =
+	        MmeCommonUtils::findControlBlockForS11Msg(eMsg->getMsgBuffer());
 	if(controlBlk_p == NULL)
 	{
-		log_msg(LOG_ERROR, "handleReleaseBearerResponse_v: "
+		log_msg(LOG_ERROR, "handleDdnMsg_v: "
```
Comment (thakurajayL) [inline src/mme-app/msgHandlers/gtpMsgHandler.cpp:179]: add enhancement defect to reject DDN with cause UNKNOWN IMSI. This helps SGW to cleanup the session.

File: src/s11/handlers/ddn_ack_handler.c [hunk 4]
```diff
@@ -59,7 +59,7 @@ ddn_ack_processing(struct DDN_ACK_Q_msg *ddn_ack_msg)
 	gtpHeader.msgType =  GTP_DOWNLINK_DATA_NOTIFICATION_ACK;
 	gtpHeader.sequenceNumber = ddn_ack_msg->seq_no;
 	gtpHeader.teidPresent = true;
-	gtpHeader.teid = ddn_ack_msg->ue_idx;
+	gtpHeader.teid = ddn_ack_msg->s11_sgw_cp_teid;
 
 	g_s11_sequence++;
```
Comment (thakurajayL) [inline src/s11/handlers/ddn_ack_handler.c:64]: We need sequence management API/module within GTP. Why to increase sequence number while sending Ack message ?

Comment (clstefi) [reply src/s11/handlers/ddn_ack_handler.c]: Sequence number need not be incremented. Addressed in the commit https://github.com/omec-project/Nucleus/pull/15/commits/9d5ceca6255e143e79cf646ed94153e5fa0a37bf

File: src/mme-app/actionHandlers/attachActionHandlers.cpp (hunk 1 above)
Comment (asreek) [reply src/mme-app/actionHandlers/attachActionHandlers.cpp:658]: Opened EDGEPOD-420

File: src/mme-app/actionHandlers/serviceRequestActionHandlers.cpp (hunk 2 above)
Comment (asreek) [reply src/mme-app/actionHandlers/serviceRequestActionHandlers.cpp:136]: Opened EDGEPOD-421.

File: src/mme-app/msgHandlers/gtpMsgHandler.cpp (hunk 3 above)
Comment (asreek) [reply src/mme-app/msgHandlers/gtpMsgHandler.cpp:179]: Here Control block will be NULL only if ue index in DDN message is 0. All other cases will be handled in default ddn handler, where we will respond with 'ue context not found' cause.

**Comment by thakurajayL** [comment]
+ retest this please

**Pull request merged**